Port login server to rust

easong-openai opened the pull request:
Port the login server to rust.

easong-openai committed:
initial

easong-openai committed:
cleanup

pakrym-oai commented on codex-rs/login/src/token_data.rs:
```diff
@@ -129,12 +129,54 @@ pub(crate) fn parse_id_token(id_token: &str) -> Result<IdTokenInfo, IdTokenInfoE
+/// deserializer used by `TokenData.id_token`.
+///
+/// Why accept both a raw JWT string and an object?
+/// - Python flow (historical): The original Python login helper wrote
```
is there a reason not to keep doing this?

pakrym-oai commented on codex-rs/login/src/server.rs:
```diff
@@ -0,0 +1,452 @@
+    let mut parts = jwt.split('.');
+    let (_h, payload_b64, _s) = match (parts.next(), parts.next(), parts.next()) {
+        (Some(h), Some(p), Some(s)) if !h.is_empty() && !p.is_empty() && !s.is_empty() => (h, p, s),
+        _ => return serde_json::Map::new(),
```
should we return an error/log here and below

easong-openai committed:
error handling, remove old script

easong-openai committed:
comment

easong-openai committed:
fmt

pakrym-oai committed:
delete some lines

bolinfest commented on codex-rs/login/src/assets/success.html:
```diff
@@ -0,0 +1,199 @@
+    </script>
+  </body>
+  </html>
+
```
strip extra blank lines?

bolinfest commented on codex-rs/login/src/bin/codex-login-server.rs:
```diff
@@ -0,0 +1,29 @@
+use std::path::PathBuf;
```
Is this just for testing? I don't see a `[bin]` entry in `Cargo.toml`.

pakrym-oai replied:
removed

bolinfest commented on codex-rs/login/tests/login_server_e2e.rs:
```diff
@@ -0,0 +1,175 @@
+
+#[test]
+fn end_to_end_login_flow_persists_auth_json() {
+    let (issuer_addr, issuer_handle) = start_mock_issuer();
```
I suspect these tests need our standard:

```
    if std::env::var(CODEX_SANDBOX_NETWORK_DISABLED_ENV_VAR).is_ok() {
        println!(
            "Skipping test because it cannot execute when network is disabled in a Codex sandbox."
        );
        return;
    }
```

We should probably create a macro for that...

pakrym-oai replied:
Added

bolinfest commented on codex-rs/login/src/lib.rs:
```diff
@@ -258,133 +256,72 @@ pub fn logout(codex_home: &Path) -> std::io::Result<bool> {
-    pub child: Arc<Mutex<Child>>,
-    pub stdout: Arc<Mutex<Vec<u8>>>,
-    pub stderr: Arc<Mutex<Vec<u8>>>,
+    url: Arc<Mutex<Option<String>>>,
```
The docstring needs to be updated.

pakrym-oai replied:
Updated

bolinfest replied:
`cancel()` is on `SpawnedLogin`, not a particular mutex, correct?

pakrym-oai replied:
updated

pakrym-oai committed:
login: refactor id_token serialization/deserialization

- Use a custom serializer for TokenData.id_token to always emit the raw JWT string rather than the expanded object form.
- Track the raw JWT as IdTokenInfo.raw_jwt for roundtripping.
- Simplify deserialize_id_token to only accept string-form JWTs (removing object compatibility).
- Update tests for roundtrip verification of auth.json using string-only id_token.
- Remove obsolete backcompat_auth_json.rs test for object form support.

pakrym-oai committed:
login: fix typo in server port error message

easong-openai committed:
onboarding, remove login server bin

bolinfest commented on codex-rs/login/src/lib.rs:
```diff
@@ -258,133 +256,72 @@ pub fn logout(codex_home: &Path) -> std::io::Result<bool> {
-            b.extend_from_slice(data);
-        }
-        Ok(data.len())
+    pub fn try_status(&self) -> Option<bool> {
```
What does this method name mean?

pakrym-oai replied:
renamed

bolinfest commented on codex-rs/login/src/lib.rs:
```diff
@@ -258,133 +256,72 @@ pub fn logout(codex_home: &Path) -> std::io::Result<bool> {
-    fn flush(&mut self) -> io::Result<()> {
-        Ok(())
+    pub fn cancel(&self) {
+        self.shutdown
```
Should we use `tokio::sync::Notify` instead?

easong-openai committed:
merge

bolinfest commented on codex-rs/login/src/lib.rs:
```diff
@@ -258,133 +256,72 @@ pub fn logout(codex_home: &Path) -> std::io::Result<bool> {
-
-    let (_file, path) = tmp.keep()?;
-    Ok(path)
+pub async fn login_with_chatgpt(codex_home: &Path) -> std::io::Result<()> {
```
Is this the main entry point to this function? If so, can this be closer to the top of the file? And we have a docstring that explains what the contract of this is? If it returns `Ok`, does that mean that `auth.json` is written?

pakrym-oai replied:
Moved to the top of login.rs

pakrym-oai committed:
cli/login: move login_with_chatgpt to cli crate; print local login server URL

Relocated the async `login_with_chatgpt` function from codex-login into the CLI crate, updating it to print the browser auth URL and port when starting the local login server. Removed the lib.rs version. Uses mpsc to notify the user of the URL.

bolinfest reviewed (commented):
Bear in mind that we will want to provide MCP requests around this business logic. We don't have to solve that in this PR, but I want to make sure we're set up well for it.

pakrym-oai committed:
fb

bolinfest commented on codex-rs/cli/src/login.rs:
```diff
@@ -1,20 +1,55 @@
+    .map_err(std::io::Error::other)??;
+
+    eprintln!("Successfully logged in");
+    drop(url_printer);
```
Why the explicit `drop()`?

bolinfest commented on codex-rs/cli/src/login.rs:
```diff
@@ -1,20 +1,55 @@
+
+    tokio::task::spawn_blocking(move || {
+        let opts = ServerOptions::new(&codex_home, client_id);
+        run_server_blocking_with_notify(opts, Some(tx), None)
```
I expected this function to share more logic from the `codex-login` crate directly, but I guess this is the main thing?

bolinfest commented on codex-rs/login/src/lib.rs:
```diff
@@ -255,138 +253,65 @@ pub fn logout(codex_home: &Path) -> std::io::Result<bool> {
-            b.extend_from_slice(data);
-        }
-        Ok(data.len())
+    pub fn get_auth_result(&self) -> Option<bool> {
```
It's not obvious how to use `SpawnedLogin`. I have to piece it together from how it is used in `codex-rs/tui/src/onboarding/auth.rs`.

In particular, it seems like this function is meant to be polled until it returns `Some`: is that correct?

pakrym-oai replied:
Yeah, it would be better if it was async and we can wait for server to complete but UI code it's invoked from isn't async.

bolinfest commented on codex-rs/login/src/lib.rs:
```diff
@@ -255,138 +253,65 @@ pub fn logout(codex_home: &Path) -> std::io::Result<bool> {
-                    .next_back()
-                    .map(|s| s.to_string())
-            })
+        self.url.lock().ok().and_then(|u| u.clone())
```
It's not clear if/when it is safe to invoke this method.

pakrym-oai committed:
cli/login: remove unused tokio::spawn handle and drop in login_with_chatgpt

login: clarify SpawnedLogin doc; reference stopping the login server with cancel()
